linera-indexer/src/lib.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module defines the Linera indexer.
+
+pub mod application;
+pub mod applications;
+pub mod graphql;
+pub mod messages;
+pub mod operations;
+pub mod plugin;
+pub mod types;
+
+use crate::{
+    graphql::{block, Block},
+    types::IndexerError,
+};
+use async_recursion::async_recursion;
+use graphql_client::reqwest::post_graphql;
+use linera_base::{crypto::CryptoHash, data_types::BlockHeight, identifiers::ChainId};
+use linera_chain::data_types::HashedValue;
+use linera_views::{
+    common::{ContextFromDb, KeyValueStoreClient},
+    map_view::MapView,
+    value_splitting::DatabaseConsistencyError,
+    views::{RootView, View, ViewError},
+};
+use operations::OperationsPlugin;
+use std::{cmp::Ordering, sync::Arc};
+use tokio::sync::Mutex;
+use tracing::info;
+
+#[derive(RootView)]
+struct StateView<C> {
+    chains: MapView<C, ChainId, CryptoHash>,
+}
+
+pub struct Indexer<C> {
+    start: BlockHeight,
+    node: String,
+    state: Arc<Mutex<StateView<C>>>,
+    pub operations: Option<OperationsPlugin<C>>,
+}
+
+impl<DB> Indexer<ContextFromDb<(), DB>>
+where
+    DB: KeyValueStoreClient + Clone + Send + Sync + 'static,
+    DB::Error: From<bcs::Error>
+        + From<DatabaseConsistencyError>
+        + Send
+        + Sync
+        + std::error::Error
+        + 'static,
+    ViewError: From<DB::Error>,
+{
+    /// Loads all indexer plugins
+    pub async fn load(
+        plugins: &[&str],
+        client: DB,
+        start: BlockHeight,
+        node: String,
+    ) -> Result<Self, IndexerError> {
+        let operations = if plugins.contains(&"operations") {
+            let context = ContextFromDb::create(client.clone(), vec![1], ())
```

Comment:
We should probably have constants for the prefixes. Ideally all in the same place so that we don't accidentally create conflicting ones.

Reply:
would it be a good idea to take the hash (or a part of the hash) of the name as a prefix?

Reply:
Maybe, sounds good to me.
@MathieuDutSik: Do you have a recommendation for choosing prefixes in cases like this?

Reply:
The choice of prefixes is not really important but has to be done so that contexts do not overlap.
Some of the methods used for achieving that are detailed in `linera-views/DESIGN.md`. Let me know if that need expansion.
I would not use the hash of a name as a prefix since it exposes to collisions (I know unlikely but still) and because it cannot be interpreted back. Instead, I would serialize the string as a `Vec<u8>` and append this to the preceding prefix to get a new prefix.